sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/batch/BulkExecutor.java

```diff
@@ -156,111 +159,172 @@ public BulkExecutor(CosmosAsyncContainer container,
 
     public Flux<CosmosBulkOperationResponse<TContext>> execute() {
 
-        return this.inputOperations
-            .onErrorContinue((throwable, o) ->
-                logger.error("Skipping an error operation while processing {}. Cause: {}, Context: {}",
-                    o,
-                    throwable.getMessage(),
-                    this.operationContextText))
-            .doOnNext((CosmosItemOperation cosmosItemOperation) -> {
-
-                // Set the retry policy before starting execution. Should only happens once.
-                BulkExecutorUtil.setRetryPolicyForBulk(
-                    docClientWrapper,
-                    this.container,
-                    cosmosItemOperation,
-                    this.throttlingRetryOptions);
-
-                if (cosmosItemOperation != FlushBuffersItemOperation.singleton()) {
-                    totalCount.incrementAndGet();
-                }
-            })
-            .doOnComplete(() -> {
-                mainSourceCompleted.set(true);
-
-                long totalCountSnapshot = totalCount.get();
-                logger.debug("Main source completed - # left items {}, Context: {}",
-                    totalCountSnapshot,
+        // The groupBy below is running into a hng if the flatMap above is
+        // not allowing at least a concurrency of the number of unique values
+        // you groupBy on.
+        // The groupBy is used to isolate Cosmos physical partitions
+        // so when there is no config override we enforce that the flatMap is using a concurrency of
+        // Math.max(default concurrency (256), #of partitions * 2 (to accommodate for some splits))
+        // The config override can beu sed by the Spark connector when customers follow best practices and
+        // repartition teh data frame to avoid that each Spark partition contains data spread across all
+        // physical partitions. When repartitioning the incoming data it is possible to ensure that each
+        // Spark partition will only target a subset of Cosmos partitions. This will improve the efficiency
+        // and mean fewer than #of Partitions concurrency will be needed for
+        // large containers. (with hundreds of physical partitions)
+        Integer nullableMaxConcurrentCosmosPartitions = ImplementationBridgeHelpers.CosmosBulkExecutionOptionsHelper
+            .getCosmosBulkExecutionOptionsAccessor()
+            .getMaxConcurrentCosmosPartitions(cosmosBulkExecutionOptions);
+        Mono<Integer> maxConcurrentCosmosPartitionsMono = nullableMaxConcurrentCosmosPartitions != null ?
+            Mono.just(Math.max(256, nullableMaxConcurrentCosmosPartitions)) :
+            this.container.getFeedRanges().map(ranges -> Math.max(256, ranges.size() * 2));
```

Comment:
by theory the number of the physical partitions can keep growing? 
wonder the perf about a container with 1000 partitions or more.

feel like long long term, we should consider remove groupby.